src/main/java/io/vertx/core/http/impl/pool/Pool.java

```diff
@@ -0,0 +1,315 @@
+ * ------------------------------------------------------
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * and Apache License v2.0 which accompanies this distribution.
+ *
+ *     The Eclipse Public License is available at
+ *     http://www.eclipse.org/legal/epl-v10.html
+ *
+ *     The Apache License v2.0 is available at
+ *     http://www.opensource.org/licenses/apache2.0.php
+ *
+ * You may elect to redistribute this code under either of these licenses.
+ */
+package io.vertx.core.http.impl.pool;
+
+import io.netty.channel.Channel;
+import io.vertx.core.Handler;
+import io.vertx.core.http.ConnectionPoolTooBusyException;
+import io.vertx.core.impl.ContextImpl;
+import io.vertx.core.logging.Logger;
+import io.vertx.core.logging.LoggerFactory;
+import io.vertx.core.spi.metrics.HttpClientMetrics;
+
+import java.util.*;
+import java.util.function.BiConsumer;
+
+/**
+ * The pool is a queue of waiters and a list of connections.
+ *
+ * The pools invariant:
+ * - the pool {@link #capacity} is the sum of the {@link Holder#capacity} of the {@link #available} list
+ * - a connection in the {@link #available} list has its {@code Holder#capacity > 0}
+ * - the {@link #weight} is the sum of all {@link Holder#weight}, i.e the connections in the {@link #all} set
+ *
+ * A connection is delivered to a {@link Waiter} on the connection's event loop thread, the waiter must take care of
+ * calling {@link io.vertx.core.impl.ContextInternal#executeFromIO} if necessary.
+ *
+ * Calls to the pool are synchronized on the pool to avoid race conditions and maintain its invariants. This pool can
+ * be called from different threads safely (although it is not encouraged for performance reasons), since it synchronizes
+ * on the pool and callbacks are asynchronous on the event loop thread of the connection without holding the pool
+ * lock to avoid deadlocks. We benefit from biased locking which makes the overhead of synchronized near zero.
+ *
+ * To constrain the number of connections the pool maintains a {@link #weight} value that must remain below the the
+ * {@link #maxWeight} value to create a connection. Weight is used instead of counting connection because this pool
+ * can mix connections with different concurrency (HTTP/1 and HTTP/2) and this flexibility is necessary.
+ *
+ * When a connection is created an initial weight is returned by the {@link ConnectionProvider#connect} method and is
+ * added to the current weight. When the channel is connected the {@link ConnectionListener#onConnectSuccess} callback
+ * provides the initial weight returned by the connect method and the actual connection weight so it can be used to
+ * correct the current weight. When the channel fails to connect the {@link ConnectionListener#onConnectFailure} failure
+ * provides the initial weight so it can be used to correct the current weight.
+ *
```

Comment:
Could you add a section about what happens to idle connections?

Reply:
Could you also add a section about the pool getting closed whenever the last open connections closes? This behavior comes from the `ConnectionManager` implementation (with a map of endpoints) but if other Vert.x projects use this pool they must be aware of it.

Reply:
ok

Reply:
👍